cmd/metal-api/internal/datastore/machine.go

```diff
@@ -442,19 +436,40 @@ func (rs *RethinkStore) FindWaitingMachine(partitionid, sizeid string) (*metal.M
 		"state": map[string]string{
 			"value": string(metal.AvailableState),
 		},
-		"waiting":    true,
-		"liveliness": metal.MachineLivelinessAlive,
-	}).Sample(1)
+		"waiting": true,
+	})
+
+	var candidates metal.Machines
+	err := rs.searchEntities(&q, &candidates)
+	if err != nil {
+		return nil, err
+	}
 
-	var availables metal.Machines
-	err := rs.searchEntities(&q, &availables)
+	ecs, err := rs.ListProvisioningEventContainers()
 	if err != nil {
 		return nil, err
 	}
+	ecMap := ecs.ByID()
+
+	var available metal.Machines
+	for _, m := range candidates {
+		ec, ok := ecMap[m.ID]
+		if !ok {
+			rs.SugaredLogger.Errorw("cannot find machine provisioning event container", "machine", m, "error", err)
+			// fall through, so the rest of the machines is getting evaluated
+			continue
+		}
+		switch ec.Liveliness {
+		case metal.MachineLivelinessAlive:
+			available = append(available, m)
+		}
+	}
```

Comment:
check whether we can get the same behavior with a table join?

Reply:
We add a hint that this might be a optimization